### ISSUE TITLE
feature_request: has_variable term classifier

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#9301]
After reflecting on the discussion in #7670, I agree that the variety of classification functions would be best supported by a working tactic language to write them in, but there is a single distinguished one that should be provided in a standard library. And it also seems that it would require more effort than just tactic language support to implement.

One "tactic" that commonly needs to be guarded by a term classification check is Coq computation, mostly `cbv`/`lazy`/`vm_compute`. I believe that the correct guard would have the property that if it succeeds on a term of type `T`, then any function of type `T -> bool` applied to that term would evaluate to `true` or `false` under any normalizing evaluation strategy. I think this means that this classifier would need to ensure that the `Print Assumptions` list is empty, i.e. the term does not contain axioms or context variables. Of course this check is incomplete, but I believe this is the best we can do in linear time. Let's call it `has_variables`.

Given this, a another very useful classifier would be reasonably approximatable in ltac1. While the previous asks whether computing a boolean would succeed, this one would ask whether it would succeed "fast", i.e. whether any time would be spent computing the input of the boolean function. I believe the appropriate test here is `not (has_variables e) and not (has_redex e)`, where `has_redex` can be defined in terms of a finite number of calls to `cbv` with flags and Ltac matching. I think this is what is meant by `is_ground` above. From this, even more restrictive classifiers (e.g. head is a constructor) can be derived using ltac matching (and even more using ltac context matching).

It seems to me that the `has_variables` here does not have a nice and fast implementation in ltac1 or ltac2, but could probably be done efficiently with some cachine in ML.

**Comment by JasonGross** [comment]
> I believe that the correct guard would have the property that if it succeeds on a term of type `T`, then any function of type `T -> bool` applied to that term would evaluate to `true` or `false` under any normalizing evaluation strategy.

I think this condition needs refining.  It currently has exactly one way to satisfy it:
```coq
Ltac is_good v := fail.
```
because the function of type `T -> bool` might itself be an axiom


> It seems to me that the `has_variables` here does not have a nice and fast implementation in ltac1 or ltac2, but could probably be done efficiently with some caching in ML.

Note that `Print Assumptions` can often take over 30 seconds.

**Comment by andres-erbsen** [comment]
Indeed, the spec only needs to hold for bool-returning functions that themselves compute; having no assumptions is still the right check. I know it is currently slow; I don't think there is any reason it should be.

**Comment by SkySkimmer** [comment]
We could have a reduction flag which makes it raise an exception if it encounters a neutral (axiom/Qed/variable).
That's not quite the same spec since reducing `fst (true, some_axiom)` would succeed but that may be fine for your use case?
